calyx-opt/src/passes/simplify_static_guards.rs

```diff
@@ -0,0 +1,141 @@
+                    // one of g1 or g2 is made up entirely of static timing guards
+                    (None, Some(g)) | (Some(g), None) => Some(g),
+                    // both g1 and g2 have non-static timing guards
+                    (Some(g1_unwrapped), Some(g2_unwrapped)) => {
+                        Some(ir::Guard::And(
+                            Box::new(g1_unwrapped),
+                            Box::new(g2_unwrapped),
+                        ))
+                    }
+                };
+                // should extend g1 intervals to include g2 intervals
+                intervals_g1.extend(intervals_g2);
+                (remaining_guard, intervals_g1)
+            }
+            ir::Guard::Info(static_timing_info) => {
+                // no "rest of guard" for static intervals
+                (None, vec![static_timing_info.get_interval()])
+            }
+        }
+    }
+
+    /// Takes in a guard and returns the simplified guard
+    /// In particular if g = g1 & g2 & ...gn, then it takes all of the g_i's that
+    /// are "static timing intervals", e.g., %[2:3], and combines them into one
+    /// timing interval.
+    /// For example: (port.out | !port1.out) & (port2.out == port3.out) & %[2:8] & %[5:10] ?
+    /// becomes (port.out | !port1.out) & (port2.out == port3.out) & %[5:8] ?
+    /// by "combining: %[2:8] & %[5:10]
+    fn simplify_guard(
+        guard: ir::Guard<ir::StaticTiming>,
+    ) -> ir::Guard<ir::StaticTiming> {
+        // get the rest of the guard and the "anded intervals"
+        let (rest_guard, mut anded_intervals) =
+            Self::separate_anded_intervals(guard);
+        // first simplify the vec of `anded_intervals` into a single interval
+        let replacing_interval = {
+            if anded_intervals.is_empty() {
+                // if there were no static timing guards (i.e., no %[2:3]), then
+                // there is no "replacing intervals"
+                None
+            } else {
+                // the replacing intervals should just be the latest beginning interval
+                // combined with the earliest ending interval, since we know that all of
+                // the intervals are connected by &.
+                let (mut max_beg, mut min_end) = anded_intervals.pop().unwrap();
+                for (cur_beg, cur_end) in anded_intervals {
+                    max_beg = std::cmp::max(cur_beg, max_beg);
+                    min_end = std::cmp::min(cur_end, min_end);
+                }
+                if max_beg >= min_end {
+                    // if the vec was something like %[2:3] | %[4:5], then this is always false
+                    // if max_beg >= min_end, then guard is always false
+                    return ir::Guard::Not(Box::new(ir::Guard::True));
+                } else {
+                    // otherwise return the single interval as the "new" interval
+                    Some(ir::Guard::Info(ir::StaticTiming::new((
```

Comment:
This makes me think that `ir::StaticTiming` should panic if the start is greater than the end just for sanity checking